server/src/transport/ws.rs

```diff
@@ -274,62 +286,104 @@ enum Receive<S> {
-	ping_config: Option<PingConfig>,
-	missed_pings: &mut usize,
-) -> Receive<S>
+async fn try_recv<T, S>(ws_stream: &mut T, mut stopped: S, ping_tx: &mpsc::Sender<KeepAlive>) -> Receive<S>
 where
 	S: Future<Output = ()> + Unpin,
 	T: StreamExt<Item = Result<Incoming, SokettoError>> + Unpin,
 {
-	let mut last_active = Instant::now();
-	let inactivity_check = match ping_config {
-		Some(p) => IntervalStream::new(interval_at(tokio::time::Instant::now() + p.ping_interval, p.ping_interval)),
-		None => IntervalStream::pending(),
-	};
-
-	tokio::pin!(inactivity_check);
-
-	let mut futs = futures_util::future::select(ws_stream.next(), inactivity_check.next());
-
 	loop {
-		match futures_util::future::select(futs, stopped).await {
+		match futures_util::future::select(ws_stream.next(), stopped).await {
 			// The connection is closed.
-			Either::Left((Either::Left((None, _)), _)) => break Receive::ConnectionClosed,
+			Either::Left((None, _)) => break Receive::ConnectionClosed,
 			// The message has been received, we are done
-			Either::Left((Either::Left((Some(Ok(Incoming::Data(d))), _)), s)) => break Receive::Ok(d, s),
+			Either::Left((Some(Ok(Incoming::Data(d))), s)) => {
+				let ping_tx = ping_tx.clone();
+				tokio::spawn(async move {
+					_ = ping_tx.send(KeepAlive::Pong(Instant::now())).await;
+				});
+
+				break Receive::Ok(d, s);
+			}
 			// Got a pong response, update our "last seen" timestamp.
-			Either::Left((Either::Left((Some(Ok(Incoming::Pong)), inactive)), s)) => {
-				last_active = Instant::now();
+			Either::Left((Some(Ok(Incoming::Pong)), s)) => {
+				let ping_tx = ping_tx.clone();
+				tokio::spawn(async move {
+					_ = ping_tx.send(KeepAlive::Pong(Instant::now())).await;
+				});
+
 				stopped = s;
-				futs = futures_util::future::select(ws_stream.next(), inactive);
 			}
 			// Received an error, terminate the connection.
-			Either::Left((Either::Left((Some(Err(e)), _)), s)) => break Receive::Err(e, s),
-			// Max inactivity timeout fired, check if the connection has been idle too long.
-			Either::Left((Either::Right((_instant, rcv)), s)) => {
-				if let Some(p) = ping_config {
-					if last_active.elapsed() > p.inactive_limit {
-						*missed_pings += 1;
-
-						if *missed_pings >= p.max_failures {
-							tracing::debug!(
-								target: LOG_TARGET,
-								"WS ping/pong inactivity limit `{}` exceeded; closing connection",
-								p.max_failures,
-							);
-							break Receive::ConnectionClosed;
-						}
+			Either::Left((Some(Err(e)), s)) => break Receive::Err(e, s),
+			// Server has been stopped.
+			Either::Right(_) => break Receive::Stopped,
+		}
+	}
+}
+
+#[derive(Debug, Copy, Clone)]
+enum KeepAlive {
+	Ping(Instant),
+	Pong(Instant),
+}
+
+async fn ping_pong_task(mut rx: mpsc::Receiver<KeepAlive>, max_inactive_limit: Duration, max_inactive: usize) {
+	let mut polling_interval = IntervalStream::new(interval(max_inactive_limit));
+	let mut pending_pings: VecDeque<Instant> = VecDeque::new();
```

Comment:
The only annoyance with this approach is that this may become big if the max_inactive_limit is way bigger than the ping interval however that's configurable thingy and I don't think it's a real issue as we also regard received messages as pong.